upgrades/tests/test_cluster.py

```diff
@@ -0,0 +1,163 @@
+def test_WHEN_wait_for_cluster_to_start_up_AND_max_wait_exceeded_THEN_raises(mock_time):
+    # Set up test
+    test_cluster_name = "cluster-name"
+    test_num_nodes = 2
+    mock_docker_client = mock.Mock()
+    test_image = "image"
+
+    mock_node_1 = mock.Mock()
+    mock_node_1.is_active.side_effect = [False, False, True]
+    mock_node_2 = mock.Mock()
+    mock_node_2.is_active.side_effect = [False, True]
+
+    test_cluster = cluster.Cluster(test_cluster_name, test_num_nodes, mock_docker_client, test_image)
+    test_cluster._nodes = {
+        test_cluster._generate_node_name(1): mock_node_1,
+        test_cluster._generate_node_name(2): mock_node_2
+    }
+
+    # Run our test
+    with pytest.raises(cluster.ClusterNotStartedInTimeException):
+        test_cluster.wait_for_cluster_to_start_up(1)
+
+    # Check the results
+    assert 1 == mock_time.sleep.call_count
+    assert 2 == mock_node_1.is_active.call_count
+    assert 2 == mock_node_2.is_active.call_count
+
+def test_WHEN_stop_THEN_as_expected():
+    # Set up test
+    test_cluster_name = "cluster-name"
+    test_num_nodes = 2
+    mock_docker_client = mock.Mock()
+    test_image = "image"
+
+    mock_node_1 = mock.Mock()
+    mock_node_1.is_active.side_effect = [False, False, True]
+    mock_node_2 = mock.Mock()
+    mock_node_2.is_active.side_effect = [False, True]
+
+    test_cluster = cluster.Cluster(test_cluster_name, test_num_nodes, mock_docker_client, test_image)
+    test_cluster._nodes = {
+        test_cluster._generate_node_name(1): mock_node_1,
+        test_cluster._generate_node_name(2): mock_node_2
+        }
+
+    # Run our test
+    test_cluster.stop()
+
+    # Check the results
+    assert 1 == mock_node_1.stop.call_count
+    assert 1 == mock_node_2.stop.call_count
+
+def test_WHEN_clean_up_THEN_as_expected():
+    # Set up test
+    test_cluster_name = "cluster-name"
+    test_num_nodes = 2
+    mock_docker_client = mock.Mock()
+    test_image = "image"
+
+    mock_network = mock.Mock()
+    mock_volume_1 = mock.Mock()
+    mock_volume_2 = mock.Mock()
+    mock_node_1 = mock.Mock()
+    mock_node_2 = mock.Mock()
+
+    test_cluster = cluster.Cluster(test_cluster_name, test_num_nodes, mock_docker_client, test_image)
+    test_cluster._networks = [mock_network]
+    test_cluster._volumes = [mock_volume_1, mock_volume_2]
+    test_cluster._nodes = {
+        test_cluster._generate_node_name(1): mock_node_1,
+        test_cluster._generate_node_name(2): mock_node_2
+        }
+
+    # Run our test
+    test_cluster.clean_up()
+
+    # Check the results
+    assert 1 == mock_node_1.clean_up.call_count
+    assert 1 == mock_node_2.clean_up.call_count
+
+    expected_remove_networks = [mock.call(mock_network)]
+    assert expected_remove_networks == test_cluster._docker_client.remove_network.call_args_list
+
+    expected_remove_networks = [mock.call(mock_volume_1), mock.call(mock_volume_2)]
```

Comment:
I think this should be `expected_remove_volumes`

Reply:
Ah, yep - thanks!

Reply:
Addressed in next rev